### PR TITLE
[BACK] JWT and Passport Integration

### PR DESCRIPTION
**Pull request by clinche**
JWT implmentation and authentication should now be complete on nestjs. All routes are protected by default, and the token is automatically verified. A `@Public` decorator has been made to exclude certain routes from protection

To test this pull request, go to `localhost:3000/auth/profile`, it should return 401. 
Then, go to `localhost:4200/login`, open network inspector and click login. you will see a failed request to `localhost:3000` (idk why but i don't think its important), open it in a new tab and it should succeed by returning you an access_token.
Next, do a GET to `localhost:3000/auth/profile` with the `Authorization` header set as follows: `Authorization: Bearer {ACCESS_TOKEN}`, it should succeeds

What it has done (or should has done) is get your code, talk to 42api to get a token, get your login with it, check you're in the db, and then:
- you're in the db, you get an access_token with your user_id
- you're not in the db, you're created and then given an access_token with you new user_id

### CODE REVIEW

**Comment by clinche** [comment]
/login now successfully passes code to backend through proxy and stores the accesstoken in localstorage

File: src/nestjs/src/app.controller.ts
Comment (Pixailz) [inline src/nestjs/src/app.controller.ts:1]: what's the `ft_callback` for ?

File: src/nestjs/src/public.decorator.ts
Comment (Pixailz) [inline src/nestjs/src/public.decorator.ts:1]: what this `MetaData` change ?

**Review by Pixailz — APPROVED** [review]
great code, it do basically what i've done, more cleanly (don'
t know how to take this.)

File: src/nestjs/src/app.controller.ts
Comment (clinche) [reply src/nestjs/src/app.controller.ts:1]: as in the future we plan to integrate various other authentications means, we need to differentiate callbacks now
a callback mechanism is to receive a code and perform final steps, in this case, fetch a token and check it is valid
we also need this route public as needed for authentication

File: src/nestjs/src/public.decorator.ts
Comment (clinche) [reply src/nestjs/src/public.decorator.ts:1]: it just sets a variable like an environment variable (in this case IS_PUBLIC_KEY set to 'isPublic' which is a boolean.
in jwt-auth.guard, we check if this variable evaluates to true and, if so, bypass authentication as the route should be public

**Comment by clinche** [comment]
> it do basically what i've done

it actually doesn't, and i think it's important to know why
here's the basic differences:
- we needed to separate api module (which talks to an api) and auth module (which handles authentications)
  - this is needed to permit the use of JWT in all code, not only in the api42oauth module
  - it anticipates various future new means of logging in
  - it tidies code and cleans imports
  - it permits to integrate auth as a default and not as a feature
- authentication is not manually needed
  - previously we needed the backend to handle cookies, which it should not be capable to do
  - now, every route of the server is protected **by default**, and we need to specify which is public
  - every request gets its `Authorization` header checked, its JWT decrypted and validated
  - the user information is also added directly to the request object (`req.user`)
  - if a auth problem happens, request are bounced without ever being executed
- client now do not know about existence of a back-end
  - by using angular dev server proxy, we emulate a prod environment in which port 3000 is unavalaible
  - it also permits us to rewrite basically any url without the need to setup an nginx server, and to create a /api route that directly maps to the backend
  - when talking to backend, we just need to retrieve the `access_token` from localStorage and append it in the backend request 

> (don't know how to take this.)

if we wrote perfect code every time issues and pull requests wouldn't be a thing.
never take it bad

**Comment by clinche** [comment]
To-do before merge: 
- patch auth module to differentiate between sign-ups
- patch front-end requests uris to remove `http://localhost:4200` part (URIs will always be in form `/api/{route}` or will user angular router)
- move back the `auth/ft_callback` controller into auth plugin (i was testing default protection of routes)

**Review by Pixailz — APPROVED** [review]
less go

**Pull request merged**